rust/arrow/src/compute/kernels/sort.rs

```diff
@@ -148,58 +206,165 @@ impl Default for SortOptions {
     }
 }
 
-/// Sort primitive values, excluding floats
+/// Sort primitive values
 fn sort_primitive<T>(
     values: &ArrayRef,
-    value_indices: Vec<usize>,
+    value_indices: Vec<u32>,
     null_indices: Vec<u32>,
+    nan_indices: Vec<u32>,
     options: &SortOptions,
 ) -> Result<UInt32Array>
 where
     T: ArrowPrimitiveType,
     T::Native: std::cmp::PartialOrd,
 {
     let values = as_primitive_array::<T>(values);
+    sort_primitive_typed(values, value_indices, null_indices, nan_indices, options)
+}
+
+fn sort_primitive_typed<T>(
```

Comment:
IMO we should document this behavior (order of `nan null value`).

I.e. the description in issue ARROW-9895 should be encapsulated in the code base, both in-line comments for the developer, documentation strings for the users in the logical plan, physical plan and kernel, and potentially also in the README.